tls/s2n_ktls.c

```diff
@@ -15,11 +15,176 @@
 
 #include "tls/s2n_ktls.h"
 
+#include <sys/socket.h>
+
+#include "utils/s2n_socket.h"
+
+/* These variables are used to disable ktls mechanisms during testing. */
+static bool disable_ktls_socket_config_for_testing = false;
+
 bool s2n_ktls_is_supported_on_platform()
 {
-#if defined(__linux__)
+#if S2N_KTLS_SUPPORTED
     return true;
 #else
     return false;
 #endif
 }
+
+S2N_RESULT s2n_ktls_validate(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    const struct s2n_cipher *cipher = conn->secure->cipher_suite->record_alg->cipher;
+    RESULT_ENSURE_REF(cipher);
+
+    /* kTLS enable should only be called once the handshake has completed. */
+    if (!is_handshake_complete(conn)) {
+        RESULT_BAIL(S2N_ERR_KTLS_HANDSHAKE_NOT_COMPLETE);
+    }
+
+    /* TODO support TLS 1.3
+     *
+     * TLS 1.3 support requires sending the KeyUpdate message when the cryptographic
+     * KeyLimits are met. However, this is currently only possible by applying a
+     * kernel patch to support this functionality.
+     */
+    RESULT_ENSURE(conn->actual_protocol_version == S2N_TLS12, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* Check if the cipher supports kTLS */
+    RESULT_ENSURE(cipher->ktls_supported, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+
+    /* kTLS I/O functionality is managed by s2n-tls. kTLS cannot be enabled if the
+     * application sets custom I/O (managed_send_io == false means application has
+     * set custom I/O).
+     */
+    switch (ktls_mode) {
+        case S2N_KTLS_MODE_SEND:
+            RESULT_ENSURE_REF(conn->send_io_context);
+            RESULT_ENSURE(conn->managed_send_io, S2N_ERR_KTLS_MANAGED_IO);
+
+            /* The output stuffer should be empty before enabling kTLS. */
+            RESULT_ENSURE(s2n_stuffer_data_available(&conn->out) == 0, S2N_ERR_RECORD_STUFFER_NEEDS_DRAINING);
```

Comment:
Interesting. So could you send/recv some application data before enabling ktls?

Reply:
Here is an example from the inet tests: https://github.com/aws/s2n-tls/blob/dae8e673be0089034e5013bb0563422e9ba4fe0a/tests/unit/s2n_self_talk_inet_socket_test.c#L51-L54

The usage pattern is going to be to negotiate a handshake and then enable kTLS. However its possible to send/recv data prior to enabling kTLS. We need to make sure that all the data has been drained since we dont want to handle kTLS and also regular IO.